src/lib/snarky/src/snark0.ml

```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
-            go (k y) handler s'
-        | Clear_handler (t, k) ->
-            let s', y = go t Request.Handler.fail s in
-            go (k y) handler s'
-        | Exists ({store; check; _}, p, k) ->
-            let s', value = Provider.run p get_value s handler in
-            let var = Typ.Store.run (store value) store_field_elt in
-            let (), () = go (check var) handler () in
-            go (k {Handle.var; value= Some value}) handler s'
-        | Next_auxiliary k -> go (k !next_auxiliary) handler s
-      in
-      go t0 Request.Handler.fail s0
-
-    let run_and_check' (type a s) (t0 : (a, s) t) (s0 : s) =
-      let next_auxiliary = ref 1 in
-      let aux = Field.Vector.create () in
-      let get_value : Cvar.t -> Field.t =
-        let get_one v = Field.Vector.get aux (Backend.Var.index v - 1) in
-        Cvar.eval get_one
+        incr next_auxiliary ; v
       in
-      let store_field_elt x =
-        let v = Backend.Var.create !next_auxiliary in
-        incr next_auxiliary ;
-        Field.Vector.emplace_back aux x ;
-        v
+      let run_as_prover x s =
+        match (x, s) with
+        | Some x, Some s ->
+            let s', y = As_prover.run x get_value s in
+            (Some s', Some y)
+        | _, _ -> (None, None)
       in
-      let system = R1CS_constraint_system.create () in
-      R1CS_constraint_system.set_primary_input_size system 0 ;
+      ignore
+        (Option.map system ~f:(fun system ->
+             R1CS_constraint_system.set_primary_input_size system num_inputs )) ;
       let rec go : type a s.
-          string list -> (a, s) t -> Request.Handler.t -> s -> s * a =
+             string list
+          -> (a, s) t
+          -> Request.Handler.t
+          -> s option
+          -> s option * a =
        fun stack t handler s ->
         match t with
         | Pure x -> (s, x)
-        | With_constraint_system (f, k) -> f system ; go stack k handler s
+        | With_constraint_system (f, k) ->
+            ignore (Option.map ~f system) ;
+            go stack k handler s
         | With_label (lab, t, k) ->
             let s', y = go (lab :: stack) t handler s in
             go stack (k y) handler s'
         | As_prover (x, k) ->
-            let s', () = As_prover.run x get_value s in
+            let s', _ = run_as_prover (Some x) s in
             go stack k handler s'
         | Add_constraint (c, t) ->
-            if not (Constraint.eval c get_value) then
-              failwithf "Constraint unsatisfied:\n%s\n%s\n"
-                (Constraint.annotation c)
-                (Constraint.stack_to_string stack)
-                () ;
-            Constraint.add ~stack c system ;
+            ignore
+              (Option.map system ~f:(fun system ->
+                   (* NOTE: If s is None, we aren't evaluating and shouldn't consider constraints here. *)
+                   if Option.is_some s && not (Constraint.eval c get_value)
```

Comment:
Even in the case that we are evaluating, we still shouldn't necessarily do this check (evaluating a constraint is somewhat costly so we'd rather avoid it when we can). Maybe consider just passing a boolean that controls whether or not you do the check?